MdeModulePkg/Library/SmmCorePerformanceLib/MmCorePerformanceLib.c

```diff
@@ -0,0 +1,1360 @@
+      if (PerfId == MODULE_LOADIMAGE_START_ID) {
+        mLoadImageCount++;
+        //
+        // Cache the offset of load image start record and use to be updated by the load image end record if needed.
+        //
+        if ((CallerIdentifier == NULL) && (Attribute == PerfEntry)) {
+          mCachedLength = mSmmBootPerformanceTable->Header.Length;
+        }
+      }
+
+      if (!PcdGetBool (PcdEdkiiFpdtStringRecordEnableOnly)) {
+        FpdtRecordPtr.GuidQwordEvent->Header.Type     = FPDT_GUID_QWORD_EVENT_TYPE;
+        FpdtRecordPtr.GuidQwordEvent->Header.Length   = sizeof (FPDT_GUID_QWORD_EVENT_RECORD);
+        FpdtRecordPtr.GuidQwordEvent->Header.Revision = FPDT_RECORD_REVISION_1;
+        FpdtRecordPtr.GuidQwordEvent->ProgressID      = PerfId;
+        FpdtRecordPtr.GuidQwordEvent->Timestamp       = TimeStamp;
+        FpdtRecordPtr.GuidQwordEvent->Qword           = mLoadImageCount;
+        CopyMem (&FpdtRecordPtr.GuidQwordEvent->Guid, &ModuleGuid, sizeof (FpdtRecordPtr.GuidQwordEvent->Guid));
+        if ((PerfId == MODULE_LOADIMAGE_END_ID) && (mCachedLength != 0)) {
+          CachedFpdtRecordPtr.RecordHeader = (EFI_ACPI_5_0_FPDT_PERFORMANCE_RECORD_HEADER *)((UINT8 *)mSmmBootPerformanceTable + mCachedLength);
+          CopyMem (&CachedFpdtRecordPtr.GuidQwordEvent->Guid, &ModuleGuid, sizeof (CachedFpdtRecordPtr.GuidQwordEvent->Guid));
+          mCachedLength = 0;
+        }
+      }
+
+      break;
+
+    case PERF_EVENTSIGNAL_START_ID:
+    case PERF_EVENTSIGNAL_END_ID:
+    case PERF_CALLBACK_START_ID:
+    case PERF_CALLBACK_END_ID:
+      if ((String == NULL) || (Guid == NULL)) {
+        return EFI_INVALID_PARAMETER;
+      }
+
+      StringPtr = String;
+      if (AsciiStrLen (String) == 0) {
+        StringPtr = "unknown name";
+      }
+
+      if (!PcdGetBool (PcdEdkiiFpdtStringRecordEnableOnly)) {
+        FpdtRecordPtr.DualGuidStringEvent->Header.Type     = FPDT_DUAL_GUID_STRING_EVENT_TYPE;
+        FpdtRecordPtr.DualGuidStringEvent->Header.Length   = sizeof (FPDT_DUAL_GUID_STRING_EVENT_RECORD);
+        FpdtRecordPtr.DualGuidStringEvent->Header.Revision = FPDT_RECORD_REVISION_1;
+        FpdtRecordPtr.DualGuidStringEvent->ProgressID      = PerfId;
+        FpdtRecordPtr.DualGuidStringEvent->Timestamp       = TimeStamp;
+        CopyMem (&FpdtRecordPtr.DualGuidStringEvent->Guid1, CallerIdentifier, sizeof (FpdtRecordPtr.DualGuidStringEvent->Guid1));
+        CopyMem (&FpdtRecordPtr.DualGuidStringEvent->Guid2, Guid, sizeof (FpdtRecordPtr.DualGuidStringEvent->Guid2));
+        CopyStringIntoPerfRecordAndUpdateLength (FpdtRecordPtr.DualGuidStringEvent->String, StringPtr, &FpdtRecordPtr.DualGuidStringEvent->Header.Length);
+      }
+
+      break;
+
+    case PERF_EVENT_ID:
+    case PERF_FUNCTION_START_ID:
+    case PERF_FUNCTION_END_ID:
+    case PERF_INMODULE_START_ID:
+    case PERF_INMODULE_END_ID:
+    case PERF_CROSSMODULE_START_ID:
+    case PERF_CROSSMODULE_END_ID:
+      // MU_CHANGE [BEGIN] - CodeQL change

```

Comment:
Any reason to keep the MU_CHANGE tags? Its derived from Smm, but its changed enough to be MM specific.

Reply:
I just kept it because those changes are still not in the upstream. I'm going to take them in my edk2 PR, but I wanted to call them out separately for now to distinguish the two sets of changes.